DataFormats/CaloRecHit/interface/MultifitComputations.h

```diff
@@ -283,6 +283,94 @@ namespace calo {
       }
     }
 
+    template <typename MatrixType1, typename MatrixType2, typename MatrixType3, typename MatrixType4>
+    EIGEN_ALWAYS_INLINE EIGEN_DEVICE_FUNC void calculateChiSq(MatrixType1 const& matrixL,
+                                                              MatrixType2 const& pulseMatrixView,
+                                                              MatrixType3 const& resultAmplitudesVector,
+                                                              MatrixType4 const& inputAmplitudesView,
+                                                              float& chi2) {
+      // FIXME: this assumes pulses are on columns and samples on rows
+      constexpr auto NPULSES = MatrixType2::ColsAtCompileTime;
+      constexpr auto NSAMPLES = MatrixType2::RowsAtCompileTime;
+
+      // replace pulseMatrixView * resultAmplitudesVector - inputAmplitudesView
+      // NOTE:
+      float accum[NSAMPLES];
+      {
+        float results[NPULSES];
+
+        // preload results and permute according to the pulse offsets /////////////// ??? this is not done in ECAL
+#pragma unroll
+        for (int counter = 0; counter < NPULSES; counter++) {
+          results[counter] = resultAmplitudesVector[counter];
+        }
+
+        // load accum
+#pragma unroll
+        for (int counter = 0; counter < NSAMPLES; counter++)
+          accum[counter] = -inputAmplitudesView(counter);
+
+        // iterate
+        for (int icol = 0; icol < NPULSES; icol++) {
+          float pm_col[NSAMPLES];
+
+          // preload a column of pulse matrix
+#pragma unroll
+          for (int counter = 0; counter < NSAMPLES; counter++)
+            pm_col[counter] = __ldg(&pulseMatrixView.coeffRef(counter, icol));
```

Comment:
@fwyzard 
do I need to add here 
#ifdef __CUDA_ARCH__
as you did in 
https://github.com/cms-sw/cmssw/blob/5bef34a0f16bb3b81f3c7306994550a22db30fba/DataFormats/CaloRecHit/interface/MultifitComputations.h#L210
?

Reply:
Ah, yes right.

`__ldg` is an intrinsic understood only by CUDA.